### DISCUSSION TITLE
Performance questions

### DISCUSSION THREAD

**Discussion by shan-mcarthur** [Bodmer/TFT_eSPI#2595] [General]
I need to get super-tight on performance of my main loop and that has lead me to consider what I am doing for TFT output.  Currently, TFT output takes a very long time and most of the calls are blocking.  This has a negative impact on the performance of my primary loop, which is a concern for me.  The first thing I am considering is how I can move to using a parallel interface.  Does this library support this, and are there particular TFTs that are high performance and recommended?  The second thing I am thinking about is how to offload the TFT to a background task so that my primary loop can run without lag.  I haven't found any examples or mentions about how to do this.  Are there issues with running in the background?  I assume that I would need to ensure that all the communication to the screen would be through one task, or get creative with semaphors to ensure that there aren't two different threads trying to communicate to the screen at the same time.  Does this apply to anything talking on the SPI bus, or just the device?  Are there any other thoughts that are important for those that are focused on high performance and responsive projects?

**Comment by shan-mcarthur** [comment]
And a related follow up question:  I noticed that the TFT displays of today are rather rudimentary from an API standpoint.  Even something as simple as clearing the screen requires the MCU to spit out a large matrix of pixels of the correct color.  I would think that the TFT display would have a command for clearing the screen so that it can be done immediately and with much less impact on the MCU and SPI bus.  When we look at video cards for computers there is a very rich API that has been very successful at delivering high performance displays.  Are there displays for microcontrollers that have a richer on-display API that we can take advantage of?

**Comment by Bodmer** [comment]
1. Calls are blocking unless DMA is used or the RP2040 when using the PIO interface.
2. Higest performance TFT interface is SSD1963 with 16 bit parallel using RP2040 (~500Mbps), with other processors the interface speed is the limiting factor (40 to 80Mbps).
3. You don't say which processor you are intending to use but there is an example for using a task for ESP32:
https://github.com/Bodmer/TJpg_Decoder/tree/master/examples/FLash_array/ESP32_Dual_Core_Flash_Jpg
Other users have implemented functions in tasks but this is not something I have done.
4. There are screens with graphics controllers (e.g. Nextion) that can be commanded to clear screen or draw primitives but these are not supported by this library. There are other libraries for these screens.
